lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -27,6 +87,211 @@
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 1
+LIBPATCH = 2
+
+CHARM_FILENAME_PREFIX = "90-juju-"
+SYSCTL_DIRECTORY = Path("/etc/sysctl.d")
+SYSCTL_FILENAME = Path(f"{SYSCTL_DIRECTORY}/95-juju-sysctl.conf")
+SYSCTL_HEADER = f"""# This config file was produced by sysctl lib v{LIBAPI}.{LIBPATCH}
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+"""
+
+
+class Error(Exception):
+    """Base class of most errors raised by this library."""
+
+    def __repr__(self):
+        """Represent the Error."""
+        return "<{}.{} {}>".format(type(self).__module__, type(self).__name__, self.args)
+
+    @property
+    def name(self):
+        """Return a string representation of the model plus class."""
+        return "<{}.{}>".format(type(self).__module__, type(self).__name__)
+
+    @property
+    def message(self):
+        """Return the message passed as an argument."""
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class Config(Dict):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self._data = self._load_data()
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> str:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def charm_filepath(self) -> Path:
+        """Name for resulting charm config file."""
+        return SYSCTL_DIRECTORY / f"{CHARM_FILENAME_PREFIX}{self.name}"
+
+    def update(self, config: Dict[str, dict]) -> None:
+        """Update sysctl config options with a desired set of config params.
+
+        Args:
+            config: dictionary with keys to update:
+        ```
+        {"vm.swappiness": {"value": 10}, ...}
+        ```
+        """
+        self._parse_config(config)
+
+        # NOTE: case where own charm calls update() more than once.
+        if self.charm_filepath.exists():
+            self._merge(add_own_charm=False)
```

Comment:
I think we need to provide better description in doc-string or change logic here, because this function is called `update` and it's not following real update pattern. I'll try to explain it with example.
Calling `update` twice as shown bellow, does not ensure that `95-juju-sysctl.conf` will have `vm.max_map_count` configured. It will be applied, but there will be no information and after reboot it will be lost.
```python
self.sysctl.update({"vm.max_map_count": 262144})
self.sysctl.update({"vm.swappiness": 50})
```
From this situation I would expect that `95-juju-sysctl.conf` to have both `vm.max_map_count=262144` and `vm.swappiness=50`.

Reply:
Yes, maybe `update` is not the correct term here. It's more a `configure` scenario, where we change everything, not adding.

Reply:
Yes, if the behaviour here is to write an entire config, I agree with @rgildein -- this name is confusing. It's kind of like `dict.update`, but does something very different. What about renaming it `configure`, as you suggest?